### PR TITLE
feat: support replay in multiple instances

### PR DESCRIPTION
**Pull request by pangdayuan1**
support replay in multiple instances

### CODE REVIEW

**Review by srotony — APPROVED** [review]

File: src/main/java/com/arextest/schedule/service/DefaultDeployedEnvironmentService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author jmo
+ * @since 2021/9/13
+ */
+@Service
+@Slf4j
+public class DefaultDeployedEnvironmentService {
```
Comment (mr3) [inline src/main/java/com/arextest/schedule/service/DefaultDeployedEnvironmentService.java:20]: Please rename to `DeployedEnvironmentService`

File: src/main/java/com/arextest/schedule/plan/PlanContext.java [hunk 2]
```diff
@@ -42,8 +43,26 @@ public AppServiceOperationDescriptor findAppServiceOperationDescriptor(String op
...
+    public List<ServiceInstance> targetActiveInstance() {
+        return getActiveInstance(AppServiceDescriptor::getTargetActiveInstanceList, true);
+    }
+
+    private List<ServiceInstance> getActiveInstance(Function<AppServiceDescriptor, List<ServiceInstance>> source, boolean targetInstance) {
+        List<ServiceInstance> instanceList = Lists.newArrayList();
+        for (AppServiceDescriptor descriptor : appServiceDescriptorList) {
+            List<ServiceInstance> activeInstanceList = targetInstance ? descriptor.getTargetActiveInstanceList() : descriptor.getSourceActiveInstanceList();
+            if (CollectionUtils.isEmpty(activeInstanceList)) {
+                break;
```
Comment (mr3) [inline src/main/java/com/arextest/schedule/plan/PlanContext.java:55]: Should use `continue`

File: src/main/java/com/arextest/schedule/sender/impl/AbstractReplaySender.java [hunk 3]
```diff
@@ -67,4 +70,12 @@ protected String encodeResponseAsString(Object responseBody) {
         }
         return null;
     }
+
+    protected ServiceInstance selectLoadBalanceInstance(String caseItemId, List<ServiceInstance> serviceInstances) {
+        if (CollectionUtils.isEmpty(serviceInstances)) {
+            return null;
+        }
+        Integer index = Math.abs(caseItemId.hashCode() % serviceInstances.size());
```
Comment (mr3) [inline src/main/java/com/arextest/schedule/sender/impl/AbstractReplaySender.java:78]: Recommended use `int`

File: src/main/java/com/arextest/schedule/service/DefaultDeployedEnvironmentService.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        if (CollectionUtils.isEmpty(this.environmentProviderList)) {
+            return Collections.emptyList();
+        }
+        for (DeploymentEnvironmentProvider provider : this.environmentProviderList) {
+            return provider.getActiveInstanceList(serviceDescriptor, env);
+        }
+        return Collections.emptyList();
+    }
+
+    public ServiceInstance getActiveInstanceEvent(AppServiceDescriptor serviceDescriptor, String host) {
```
Comment (mr3) [inline src/main/java/com/arextest/schedule/service/DefaultDeployedEnvironmentService.java:47]: Recommended rename to 'getActiveInstance'

**Review by mr3 — APPROVED** [review]

**Review by srotony — APPROVED** [review]

**Pull request merged**